OpenRA.Test/PathfinderTests.cs

```diff
@@ -0,0 +1,240 @@
+				search2 = PathSearch.FromPoint(world, mi, actor, target, from, true).Reverse();
+				stopwatch.Start();
+				path5 = pathfinder.FindBidiPath(search, search2);
+
+				stopwatch.Stop();
+				search = PathSearch.FromPoint(world, mi, actor, new CPos(0, 0), new CPos(51, 100), true);
+				search2 = PathSearch.FromPoint(world, mi, actor, new CPos(51, 100), new CPos(0, 0), true).Reverse();
+				stopwatch.Start();
+				path6 = pathfinder.FindBidiPath(search, search2);
+
+				stopwatch.Stop();
+				search = PathSearch.FromPoint(world, mi, actor, new CPos(0, 0), new CPos(49, 50), true);
+				search2 = PathSearch.FromPoint(world, mi, actor, new CPos(49, 50), new CPos(0, 0), true).Reverse();
+				stopwatch.Start();
+				path7 = pathfinder.FindBidiPath(search, search2);
+
+				stopwatch.Stop();
+				search = PathSearch.FromPoint(world, mi, actor, new CPos(127, 0), new CPos(50, 101), true);
+				search2 = PathSearch.FromPoint(world, mi, actor, new CPos(50, 101), new CPos(127, 0), true).Reverse();
+				stopwatch.Start();
+				path8 = pathfinder.FindBidiPath(search, search2);
+			}
+
+			Console.WriteLine("I took " + stopwatch.ElapsedMilliseconds + " ms with new FindBidipathfinder");
+		}
+
+		static int Est1(CPos here, CPos destination)
+		{
+			var diag = Math.Min(Math.Abs(here.X - destination.X), Math.Abs(here.Y - destination.Y));
+			var straight = Math.Abs(here.X - destination.X) + Math.Abs(here.Y - destination.Y);
+
+			// Min cost to arrive from once cell to an adjacent one
+			// (125 according to tests)
+			const int D = 100;
+
+			// According to the information link, this is the shape of the function.
+			// We just extract factors to simplify.
+			var h = D * straight + (D * Constants.Sqrt2 - 2 * D) * diag;
+
+			return (int)(h * 1.001);
+		}
+
+		static int Est2(CPos here, CPos destination)
+		{
+			var diag = Math.Min(Math.Abs(here.X - destination.X), Math.Abs(here.Y - destination.Y));
+			var straight = Math.Abs(here.X - destination.X) + Math.Abs(here.Y - destination.Y);
+
+			// HACK: this relies on fp and cell-size assumptions.
+			var h = (100 * diag * Constants.Sqrt2) + 100 * (straight - (2 * diag));
+			return (int)(h * 1.001);
+		}
+
+		/// <summary>
+		/// Tests the refactor of the default heuristic for pathFinder
+		/// </summary>
+		[Test]
+		public void EstimatorsTest()
+		{
+			Assert.AreEqual(Est1(new CPos(0, 0), new CPos(20, 30)), Est2(new CPos(0, 0), new CPos(20, 30)));
+		}
+
+		/// <summary>
+		/// Test for the future feature of path smoothing for Pathfinder
+		/// </summary>
+		[Test]
+		public void RayCastingTest()
+		{
+			// Arrange
+			var sut = new RayCaster();
+			CPos source = new CPos(1, 3);
+			CPos target = new CPos(3, 0);
+
+			// Act
+			var valid = sut.RayCast(source, target);
+
+			// Assert
+		}
+	}
+
+	public class RayCaster
+	{
+		// Algorithm obtained in http://playtechs.blogspot.co.uk/2007/03/raytracing-on-grid.html
+		public IEnumerable<CPos> RayCast(CPos source, CPos target)
```

Comment:
This will be useful for solving #5294, #7498, and maybe a few others.  That's a job for another pr, though.